processor/src/main/java/org/mapstruct/ap/internal/util/SpecificCompilerWorkarounds.java

```diff
@@ -115,4 +135,161 @@ public static TypeElement replaceTypeElementIfNecessary(Elements elementUtils, T
         }
         return element;
     }
+
+    public static TypeMirror asMemberOf(DeclaredType containing, Element element, Types typeUtils,
+                                        Elements elementUtils) {
+        try {
+            // quick and dirty, until implementation below is more sound
+            return typeUtils.asMemberOf( containing, element );
+        }
+        catch ( IllegalArgumentException e ) {
+            TypeMirror eclipseWorkaround =
+                EclipseEnvironment.asMemberOf( containing, element, typeUtils, elementUtils );
+
+            if ( null != eclipseWorkaround ) {
+                return eclipseWorkaround;
+            }
+
+            throw e;
+        }
+    }
+
+    private static final class EclipseEnvironment {
+        public static TypeMirror asMemberOf(DeclaredType containing, Element element, Types typeUtils,
+                                            Elements elementUtils) {
+            if ( typeUtils instanceof TypesImpl && containing instanceof DeclaredTypeImpl
+                && element instanceof ElementImpl ) {
+                try {
+                    ElementImpl elementImpl = (ElementImpl) element;
+
+                    Field field = TypesImpl.class.getDeclaredField( "_env" );
+                    field.setAccessible( true );
+                    BaseProcessingEnvImpl env = (BaseProcessingEnvImpl) field.get( typeUtils );
+
+                    // TODO can some of the reflection here be avoided?
+                    Method bindingGetter = TypeMirrorImpl.class.getDeclaredMethod( "binding" );
+                    bindingGetter.setAccessible( true );
+                    ReferenceBinding referenceBinding = (ReferenceBinding) bindingGetter.invoke( containing );
+
+                    MethodBinding methodBinding = (MethodBinding) elementImpl._binding;
+                    // matches in super-classes have priority
+                    MethodBinding inSuperclassHiearchy =
+                        findInSuperclassHierarchy( methodBinding, env, referenceBinding );
+                    if ( inSuperclassHiearchy != null ) {
+                        return env.getFactory().newTypeMirror( inSuperclassHiearchy );
+                    }
+
+                    // if nothing was found, traverse the interfaces and collect all candidate methods that match
+                    List<MethodBinding> candidatesFromInterfaces = new ArrayList<MethodBinding>();
+                    collectFromInterfaces(
+                        methodBinding,
+                        referenceBinding,
+                        new HashSet<ReferenceBinding>(),
+                        candidatesFromInterfaces );
+
+                    java.util.Collections.sort( candidatesFromInterfaces, new Comparator<MethodBinding>() {
```

Comment:
> The compiler itself should complain if the same type parameter is bound several times in incompatible ways. And as we are traversing the hierarchy from bottom to top, we should first encounter the most-specific binding in an inheritance line if that is the concern.

The compiler complains if there are incompatible declarations within one type or it's entire hiearchy. And for one type, there is only one matching method. But if the type itself doesn't contain the method and the superclasses don't contain the method either (checking the superclass hierarchy is enough, no need to check the interfaces there), then there can still be different declarations in the interfaces directly implemented by the type (or their super-interfaces). In that case, the most specialized variant wins. For each interface type, the most specialized variant is automatically the one that was found first - you just need to compare the variants that are found in siblings of the hierarchy. Hence the different termination conditions for the recursion in `collectFromInterfaces`. The JLS probably has some better and more complete wording of this, but you can try it out rather quickly with the example classes from the chat. That's what I did... :wink: